### ISSUE TITLE
Some APIs show unexpected results when using multiple threads

### ISSUE THREAD

**Issue by UESuperGate** [the-tcpdump-group/libpcap#1174]
# Libpcap Version
version 1.11.0,  commit ab3f6a677b.

# Environment
- Ubuntu 20.04
- Linux kernel 5.4.0-146-generic
- GCC 9.4.0

# Compile Configuration
Just default configuration

# Problem Details

Four APIs (i.e., `pcap_next_etherent`, `pcap_statustostr`, `bpf_image`, and `pcap_datalink_val_to_description_or_dlt`) which can be invoked by users show unexpected results when using them in multiple threads.

The root cause is that all of them use static variable and may return the pointer to the static variable. For multi-thread programs, the static variable will also be treated as share states. The users can receive the pointer returned, but there might be race condition problem if the APIs are invoked in parallel because there is no protection (e.g, lock) for the static variable.

For example, `pcap_datalink_val_to_description_or_dlt` in `libpcap/pcap.c` can receive an integer as `dlt` and will return the pointer to `unkbuf` if the `dlt` is a unknown number.

```c
const char *
pcap_datalink_val_to_description_or_dlt(int dlt)
{
        static char unkbuf[40];
        const char *description;

        description = pcap_datalink_val_to_description(dlt);
        if (description != NULL) {
                return description;
        } else {
                (void)snprintf(unkbuf, sizeof(unkbuf), "DLT %d", dlt);
                return unkbuf;
        }
}
```

However, if it is invoked in parallel, because of the lack of protection between threads, there might be race condition and finally lead to unexpected results.

```c
// source code of "pcap_datalink_val_to_description_or_dlt_testcase"
void thread_func(int *bi) {
    char *pe = pcap_datalink_val_to_description_or_dlt(*bi);
    printf("loc == %p\n", pe);
    printf("pe == %s\n", pe);
}

int main() {
    int dlt[2] = {20230101, 10103202};

    pthread_t pt[2];

    for (int i = 0; i < 2; i++) {
        pthread_create(pt + i, NULL, thread_func, dlt + i);
    }
    sleep(1);
}
```

```shell
$ ./pcap_datalink_val_to_description_or_dlt_testcase
loc == 0x7fca9d977bc0
pe == DLT 10103202
loc == 0x7fca9d977bc0
pe == DLT 10103202
```

# How to Reproduce

There are four APIs and we have four testcases in total.

## pcap_datalink_val_to_description_or_dlt

```c
// gcc -I./libpcap/ -L./libpcap/build -Wl,-rpath,./libpcap/build/ ./pcap_datalink_val_to_description_or_dlt_testcase.c -o ./pcap_datalink_val_to_description_or_dlt_testcase -lpcap -lpthread

#include "pcap/pcap.h"
#include "pcap/namedb.h"

#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>

void thread_func(int *bi) {
    char *pe = pcap_datalink_val_to_description_or_dlt(*bi);
    printf("loc == %p\n", pe);
    printf("pe == %s\n", pe);
}

int main() {
    int dlt[2] = {20230101, 10103202};

    pthread_t pt[2];

    for (int i = 0; i < 2; i++) {
        pthread_create(pt + i, NULL, thread_func, dlt + i);
    }
    sleep(1);
}
```

## pcap_statustostr
```c
// gcc -I./libpcap/ -L./libpcap/build -Wl,-rpath,./libpcap/build/ ./pcap_statustostr_testcase.c -o ./pcap_statustostr_testcase -lpcap -lpthread

#include "pcap/pcap.h"
#include "pcap/namedb.h"

#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>

void thread_func(int *fp) {
    char *pe = pcap_statustostr(*fp);
    printf("loc == %p\n", pe);
    printf("errno == %d, pe == %s\n", *fp, pe);
}

int main() {
    int es[15] = {22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36};

    pthread_t pt[15];

    for (int i = 0; i < 15; i++) {
        pthread_create(pt + i, NULL, thread_func, es + i);
    }
    sleep(2);
}
```

## bpf_image
```c
// gcc -I./libpcap/ -L./libpcap/build -Wl,-rpath,./libpcap/build/ ./pcap_bpf_image_testcase.c -o ./pcap_bpf_image_testcase -lpcap -lpthread
#include "pcap/pcap.h"
#include "pcap/namedb.h"

#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>

void thread_func(struct bpf_insn *bi) {
    char *pe = bpf_image(bi, bi->k);
    printf("loc == %p\n", pe);
    printf("pe == %s\n", pe);
}

int main() {
    struct bpf_insn bi[2] = {{BPF_ALU|BPF_OR|BPF_X, 0, 0, 0x10}, 
                            {BPF_ALU|BPF_RSH|BPF_X, 0, 0, 0x20}};

    pthread_t pt[2];

    for (int i = 0; i < 2; i++) {
        pthread_create(pt + i, NULL, thread_func, bi + i);
    }
    sleep(1);
}
```

## pcap_next_etherent
For this API, we need two more etherent files `f1` and `f2`.
```c
// gcc ./pcap_next_etherent_testcase.c -o ./pcap_next_etherent_testcase -lpthread -lpcap

#include "pcap/pcap.h"
#include "pcap/namedb.h"

#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>

void thread_func(FILE *fp) {
    struct pcap_etherent *pe = NULL;
    pe = pcap_next_etherent(fp);
    printf("pe == %p\n", pe);
    printf("pe->addr == 0x%x, 0x%x, 0x%x, 0x%x, 0x%x, 0x%x\n", pe->addr[0], pe->addr[1], pe->addr[2], pe->addr[3], pe->addr[4], pe->addr[5]);
    printf("pe->name == %s\n", pe->name);
}

int main() {
    FILE *fp1 = fopen("./f1", "rb");
    FILE *fp2 = fopen("./f2", "rb");

    pthread_t pt_1, pt_2;
    pthread_create(&pt_1, NULL, thread_func, fp1);
    pthread_create(&pt_2, NULL, thread_func, fp2);
    sleep(5);
}
```
Content of `f1`
```
1A:0:20:11:41:4D    fayoum
2B:1:30:12:42:5D    nubian 
3C:2:40:13:43:6D    sahara    # This is a comment
4D:3:50:14:44:7D    tenere 
```

Content of `f2`
```
8:0:20:1:40:16  alice
8:0:20:1:40:15  bob 
8:0:20:1:40:7   sahara    # This is a comment
8:0:20:1:40:14  darwin
```

**Comment by guyharris** [comment]
C makes string operations *so* convenient.  C++, at least, defines a string class that allows `auto` strings that get automatically freed when they go out of scope.

For at least some of these, making the static string buffer thread-local should be sufficient.

**Comment by guyharris** [comment]
That seems, at least, to fix the pcap_datalink_val_to_description_or_dlt_testcase.c test case.

I've attached a version of pcap_datalink_val_to_description_or_dlt_testcase.c that fixes some compiler warnings. (Its name ends with .c.txt because GitHub is obnoxious and requires that.)

[pcap_datalink_val_to_description_or_dlt_testcase.c.txt](https://github.com/the-tcpdump-group/libpcap/files/11405474/pcap_datalink_val_to_description_or_dlt_testcase.c.txt)

**Comment by guyharris** [comment]
Works for pcap_statustostr_testcase.c, too.  See cleaned-up pcap_statustostr_testcase.c.txt attachment.

[pcap_statustostr_testcase.c.txt](https://github.com/the-tcpdump-group/libpcap/files/11405524/pcap_statustostr_testcase.c.txt)

**Comment by guyharris** [comment]
Works for pcap_bpf_image_testcase.c, too. See cleaned-up pcap_bpf_image_testcase.c.txt attachment.

[pcap_bpf_image_testcase.c.txt](https://github.com/the-tcpdump-group/libpcap/files/11405682/pcap_bpf_image_testcase.c.txt)

**Comment by guyharris** [comment]
Works for pcap_next_etherent_testcase.c, too. See cleaned-up pcap_next_etherent_testcase.c.txt attachment.

[pcap_next_etherent_testcase.c.txt](https://github.com/the-tcpdump-group/libpcap/files/11405734/pcap_next_etherent_testcase.c.txt)

**Comment by UESuperGate** [comment]
I was thinking about using lock to resolve the problem. Obviously thread local variable is a better choice. Thanks for you kindly reply.
